First draft of encryption code.

arunabh-singh opened the pull request:
I'm not really sure where this would fit in the code. It works standalone with my test functions.

wgenemorgan committed:
Fix typo in README.md

vaage committed:
Correct Bad Int to Long Conversion

In the serializer for Time, there was a problem with
the low bits where the source integer was negative
leading to the long being negative.

The change adds a serializer for longs that should
help protect bad custom long serialization.

vaage committed:
Finish Integration With Relay

Added a system called the Timeline which allows work to be
set to run at specific points in time. This allows work to
be queued.

To allow the relay server data to merge in with the server,
a work item is set to run every minute. When it finishs, it
sets itself to run again in a minute.

When data needs to be sent to the relay, a work item is added
to the timeline so that the current connection can be closed
before sending data to the relay.

The timeline allowed for the Hub to be removed as work items
to handle connections could be added to the timeline.

Closes #4.

vaage committed:
Finish Integration With Relay

Added a system called the Timeline which allows work to be
set to run at specific points in time. This allows work to
be queued.

To allow the relay server data to merge in with the server,
a work item is set to run every minute. When it finishs, it
sets itself to run again in a minute.

When data needs to be sent to the relay, a work item is added
to the timeline so that the current connection can be closed
before sending data to the relay.

The timeline allowed for the Hub to be removed as work items
to handle connections could be added to the timeline.

Closes #4.

vaage committed:
Finish Integration With Relay

Added a system called the Timeline which allows work to be
set to run at specific points in time. This allows work to
be queued.

To allow the relay server data to merge in with the server,
a work item is set to run every minute. When it finishs, it
sets itself to run again in a minute.

When data needs to be sent to the relay, a work item is added
to the timeline so that the current connection can be closed
before sending data to the relay.

The timeline allowed for the Hub to be removed as work items
to handle connections could be added to the timeline.

Closes #4.

ruiqimao committed:
working basic sqlite abstraction layer

ruiqimao committed:
added Table abstraction and fixed style to meet Google style requirements

ruiqimao committed:
added DBObject and ways to query for them

ruiqimao committed:
removed temporary vim file

ezhuang13 committed:
Interface for compression

vaage committed:
Load Team Info into Relay

There was a TODO in the relay server to load team information
so that the relay would be able to accept read and write
requests from servers.

This change adds the ability to read a file that can contain
the team id and team secrets that it will accept.

Closes: #9

ruiqimao committed:
working database implementation

vaage committed:
Merge pull request #5 from google/relay-integration

Finish Integration With Relay

vaage committed:
Merge branch 'master' into relay-load-team-data

vaage committed:
Merge pull request #10 from google/relay-load-team-data

Load Team Info into Relay

vaage committed:
Adding missing final statements

When compiled with Java8 there were no errors but when compiled
with Java7 there were compile errors about missing final statements
for variable use inside an inner-class.

vaage committed:
Defining Travis CI Script

vaage committed:
Merge pull request #13 from google/fix-compile-errors

Adding missing final statements

vaage committed:
Merge pull request #15 from google/travis-ci-all

Defining Travis CI Script

vaage committed:
Made Uuid a class

Before Uuid was an interface, but every implementation was
the same. Every time someone made implemented Uuid, they
would have to call Uuids.complete on it to fill in the
missing functions.

Making Uuid a class solves all these problems as all the
methods will be implemented correctly and there only needs
to be one implementation.

Generators are still sepearate.

vaage committed:
Correcting import orders

vaage committed:
Reworking Test

Moved Test from common to Util.

Changed the implementation to use Date all the time and not just for
converting it to a string.

Added the first test for Time to ensure that "fromMs" and "toMs"
yielded the same values.

ezhuang13 committed:
Compression using json

ezhuang13 committed:
Ideas for where to integrate into pipeline

ruiqimao committed:
added database tests

ezhuang13 committed:
Moving working compression model

ezhuang13 committed:
Removing original CompressionEngine

vaage committed:
Merge pull request #18 from google/simplify_uuid

Changed Uuid from an interface to a class as all implementations were the same.

vaage committed:
Merge branch 'master' into move-time

vaage committed:
Update ServerFrontEnd.java

Commit:
First draft of encryption code.

ghost commented:
I think it's best to combine this with the compression code, but I'll consult with Eric first.

Tuxdude commented on src/codeu/chat/util/encryption/Encryptor.java:
```diff
@@ -0,0 +1,46 @@
+import java.security.SecureRandom;
+import java.math.BigInteger;
+
+public class Encryptor {
```
Please add javadoc to your class and at least all the public methods.

Tuxdude commented on src/codeu/chat/util/encryption/Encryptor.java:
```diff
@@ -0,0 +1,46 @@
+        newKeys();
+    }
+
+    public BigInteger encrypt(PublicKey key, String input) {
```
Please add any comments/javadoc explaining what is the operation performed by the encrypt() method.

Tuxdude replied:
It looks like you're using asymmetric encryption. Is this for encryption at rest (i.e. storage) or for encryption between 2 parties (i.e. client and server)?

Tuxdude commented on src/codeu/chat/util/encryption/PublicKey.java:
```diff
@@ -0,0 +1,13 @@
+import java.math.BigInteger;
+
+public class PublicKey {
+    BigInteger n, e;
```
Avoid single letter variable names. Make them more descriptive instead.

Tuxdude replied:
It looks like these 2 variables can be made final since they're only assigned in the constructor.

Tuxdude commented on src/codeu/chat/util/encryption/PublicKey.java:
```diff
@@ -0,0 +1,13 @@
+        e = otherE;
+    }
+
+    public BigInteger get_n() {return n;}
```
Stick with a single naming convention for all method and variable names. In this case getE and getN would be good. Expanding E and N would be even better.

Tuxdude commented on src/codeu/chat/util/encryption/EncryptorTest.java:
```diff
@@ -0,0 +1,41 @@
+import java.math.BigInteger;
+
+public class EncryptorTest {
```
Good to write tests. Yaay! Could we make them junit based? I think Ruiqi had written some junit based tests, you can check with him or just google for more info.

Tuxdude commented on src/codeu/chat/util/encryption/Encryptor.java:
```diff
@@ -0,0 +1,46 @@
+/**
```
A general comment similar to what I told Eric as well. If you're writing a class to perform encryption/decryption, it is always better, safer and easier to use any of the standard encryption libraries to perform the encrypt/decrypt operations.

You just supply the secrets, the data and let the library do the rest, instead of reinventing the wheel here. Especially for encryption, please avoid using custom cryptographic routines, as they may not be secure. :)

vaage committed:
Merge pull request #19 from google/move-time

Reworking Test

vaage committed:
Exposing params in run scripts

Before the run script had defaults for most values. This
starting causing problems when hosting the server as it
would require the run scripts to change.

Now all the paramters are passed in when calling the run
script. If any value is missing, the script exits and a
usage message is printed.

vaage committed:
Merge pull request #30 from google/update-run-scripts

Exposing params in run scripts

ezhuang13 committed:
First attempt at deploying travis

ezhuang13 committed:
First attempt at travis integration

ezhuang13 committed:
Overriding default travis build

vaage committed:
Added Optional Param for Relay Address

The java code supports an optional relay address but
it was never passed through. This change adds support
to the run_server script to pass through an optional
relay address.

vaage committed:
Merge pull request #32 from google/update-run-server

Added Optional Param for Relay Address

vaage committed:
Changing How The Relay Processes Reads

The log for reads was a lot more complicated than it needed
to be. This change makes it simplier and adds tests to verify
the reads are working.

vaage committed:
Adding a path for persistent data

To support persistent data with the server, this change
adds a command line argument that specifies where cross
run data can be stored.

ezhuang13 committed:
Adding dependencies

ezhuang13 committed:
Finalize design layout for compression of messages, updated scripts to install dependencies

ezhuang13 committed:
Update make.sh to be compatible with Linux - for specifications and Travis CI

Commit:
Verified changes to ensure project builds in a Linux environment

ruiqimao committed:
made changes as requested in pr#1

ruiqimao committed:
Merge pull request #1 from ezhuang13/database

Database

ezhuang13 committed:
Removing tests, will be added in later PR

ezhuang13 committed:
Adding merges from master, resolving merge conflicts

crepricg committed:
README update with new cmdline arguments

crepricg committed:
Adding specification for the secret server to be an hex string.

crepricg committed:
Replaces 'or' with 'and'.

vaage committed:
Merge pull request #40 from google/update_readme

README update with new cmdline arguments

crepricg committed:
Fixes HEX conversion for secret.

* Corrects the handling of uppercase letters by adding the necessary
offset.
* Handles lowercase letters as well
* Adds tests for both cases.

crepricg committed:
Merge branch 'master' into fix_hex_conversion

vaage committed:
Merge pull request #42 from google/fix_hex_conversion

Fixes HEX conversion for secret.

vaage committed:
Merge branch 'master' into relay-responding

vaage committed:
Improved messaging of variables in run_server.sh

This change improves the messaging for each variable in
that the run_server.sh script expects. Running ther script
with any value missing will yield a usage message.

vaage committed:
Merge branch 'master' into add-persistent-dir

vaage committed:
Merge pull request #33 from google/relay-responding

Changing How The Relay Processes Reads

vaage committed:
Merge branch 'master' into add-persistent-dir

vaage committed:
Update README to explain <persistent-dir>

Updated the instructions for how to call run_server.sh now that the command line arguments have changed.

vaage committed:
Merge pull request #34 from google/add-persistent-dir

Adding a path for persistent data

vaage committed:
Fix Blank Line Fail in Teams File

When the relay loads team data it tries to parse an empty
string. It should just skip empty strings.

Closes #31

vaage committed:
Merge pull request #43 from google/fix_blank_line_parse_fail

Fix Blank Line Fail in Teams File

vaage committed:
Relay re-load teams file

Before the relay would only read the teams file at the start-up. This
meant that the relay would have to be restarted whenever a new team
was added.

This change makes the relay reload the teams file every minute so that
if a new team is added, they will have access within a minute. This will
prevent the relay having to be taken down and losing all its data each
time a new team comes on line.

vaage committed:
Adding missing final statements

This fixes a compile error for Java 7 where final statements were
needed for the reload team data call.

vaage committed:
Correcting confusing comment

The comment about why the loadTeamInfo was adding itself to the timeline made no sense. This corrects the message to be comprehendible.

vaage committed:
Merge pull request #50 from google/relay-reload-teams

Relay re-load teams file

ezhuang13 committed:
Resolving merge conflicts

Commit:
Resolved merge conflicts

Commit:
Merge branch 'mergeGoogle'
Adds changes to enable RelayServer, as well as cleaning up certain aspects of the code base.

ruiqimao committed:
created server-side authentication system

ruiqimao committed:
created server-side authentication system

ruiqimao committed:
Merged changes from master

ruiqimao committed:
added authentication to server and clients

ruiqimao committed:
updated team information

ruiqimao committed:
added token checking

ruiqimao committed:
fixed formatting

ruiqimao committed:
Merge pull request #6 from ezhuang13/users

Users

Commit:
Symmetric encryption update.

ezhuang13 committed:
Compression (#7)

Implement compression of Messages, Conversations, Users, and ConversationSummaries

Commit:
Basic encryptor files.

Commit:
Cleaning up old files.

Commit:
Added basic conversation-bound symmetric encryption.

ezhuang13 committed:
Update to compression (#12)

* Reworked compression so that each class implements an interface

* Consolidating setup functions

* Small typo

Commit:
Delete эfÍظþÐú

Commit:
Delete codeu_project_2017.iml

Commit:
Delete .name

Commit:
Delete misc.xml

Commit:
Delete modules.xml

Commit:
Delete vcs.xml

Commit:
Delete workspace.xml

Commit:
Merge remote-tracking branch 'refs/remotes/origin/encryption' into encryption

Commit:
More cleaning/updating.

Commit:
Added all work to updated branch.

Commit:
Updated serializers.

Commit:
Encryptor interface and methods.

Commit:
Removed old files.

Commit:
Encryption progress.

Commit:
Working version.
